### DISCUSSION TITLE
Unique constrain exception details

### DISCUSSION THREAD

**Discussion by kujtimiihoxha** [sensedeep/dynamodb-onetable#183] [Ideas]
I do not exactly know if this even possible but I would like to have a way to know that a unique constrain has failed, this way it would be easier to respond with a meaningful message to the user.

Currently I believe the responses are generic that a DB transaction failed.

**Comment by kujtimiihoxha** [comment]
My specific usecase has to do with using `table.transact` directly instead of just `mode.create`

One approach I was thinking would be possible is using 
```
Transaction cancelled, please refer cancellation reasons for specific reasons [ConditionalCheckFailed, ConditionalCheckFailed, None, None, None, None, None, None, None]"
```
 
 In the message it tells you which one of the operations has `ConditionalCheckFailed`.
 
 and looking at the `trace` there is a way I think to parse the items in the transaction items and see if this conditional fail is because of a unique attribute.
 
 ```json
{
  "err": {
    "message": "Transaction cancelled, please refer cancellation reasons for specific reasons [ConditionalCheckFailed, ConditionalCheckFailed, None, None, None, None, None, None, None]",
    "code": "TransactionCanceledException",
    "time": "2021-11-08T14:24:02.945Z",
    "requestId": "57QUFQQVDPV9MGOE40TKDATJ2BVV4KQNSO5AEMVJF66Q9ASUAAJG",
    "statusCode": 400,
    "retryable": false,
    "retryDelay": 26.048534408821688
  },
  "trace": {
    "model": "_Generic",
    "cmd": {
      "TransactItems": [
        {
          "Put": {
            "ConditionExpression": "(attribute_not_exists(PK)) and (attribute_not_exists(SK))",
            "TableName": "my-table",
            "Item": {
              "PK": "_unique#myType#myUniqueField#value",
              "SK": "_unique#",
              "_type": "_Unique"
            },
            "ReturnValues": "NONE"
          }
        },
        {
          "Put": {
            "ConditionExpression": "(attribute_not_exists(PK)) and (attribute_not_exists(SK))",
            "TableName": "my-table",
            "Item": {
              "PK": "_unique#myType#myOtherUniqueField#value",
              "SK": "_unique#",
              "_type": "_Unique"
            },
            "ReturnValues": "NONE"
          }
        },
        ...
      ]
    }
  }
}
```
So basically you could use `errorResponse.trace.cmd.TransactItems` and see if the operation matching the `ConditionalCheckFailed` index is a `_type:'_Unique'` and than parsing the `PK` will give you which field it is 😄 .

This could be something people can implement themselves but we probably need to at least expose the trace on the error, currently it is not being exposed.

**Reply by mobsense** [reply]
Sounds valuable and feasible. 

We put the OneError in place last release to provide the platform for exactly these kinds of things. Start with more generic messages and as we see cases where warranted, add a specific message / context.

Exposing trace is a good start, but it would be better to parse and get a specific message(s). 

Do you want to try?

**Reply by kujtimiihoxha** [reply]
I will have to implement it either way for my use case so I will probably make that a PR here also :)

**Reply by kujtimiihoxha** [reply]
I had to recently implement this and just wanted to share this if someone else needs it.
```typescript
export function uniqueFieldErrorsFromContext(errContext: {
  err: AWSError;
  trace: {
    cmd: {
      TransactItems: DocumentClient.TransactWriteItemList;
    };
  };
}): Array<{ field: string; message: string }> {
  const message = errContext.err.message;

  const arrayConditionsStart = message.indexOf('[');
  const arrayConditionsEnd = message.indexOf(']');
  const conditionsOutcome = message
    .substring(arrayConditionsStart + 1, arrayConditionsEnd)
    .split(', ');

  const failedTransactions = [];

  const errorMessages = [];
  for (const index in errContext.trace.cmd.TransactItems) {
    if (conditionsOutcome[index] !== 'None') {
      failedTransactions.push(errContext.trace.cmd.TransactItems[index]);
    }
  }

  for (const transaction of failedTransactions) {
    const putItem = transaction.Put;
    if (!putItem || putItem.Item._type !== '_Unique') {
      continue;
    }
    const entity = putItem.Item.PK.split('#')[1];
    const uniqueField = putItem.Item.PK.split('#')[2];


    errorMessages.push({
      field: uniqueField,
      message: `A ${entity} with the same ${uniqueField} already exists`,
    });
  }
  return errorMessages;
}
```

Let me know if you want to embed this in the main library.

**Reply by mobsense** [reply]
Yes, please. I'd love to get a PR that improved unique messages.

I hit this issue myself recently.

Coupled with the "scope" feature to allow scoped unique items, this unique support is getting very powerful.
